tests/test_optimizer_random_dag.py

```diff
@@ -89,31 +89,36 @@ def find_min_objective(dag: sky.Dag, minimize_cost: bool) -> float:
     topo_order = dag.tasks
 
     final_plan = {}
+    min_objective = np.inf
+    resources_stack = []
 
     def _optimize_by_brute_force(tasks, plan):
         """Optimizes a Sky DAG in a brute-force manner."""
         # NOTE: Here we assume that the Sky DAG is topologically sorted.
-        nonlocal final_plan
+        nonlocal final_plan, min_objective
         task = tasks[0]
-        min_objective = np.inf
         for resources in task.get_resources():
             assert task.name in DUMMY_NODES or resources.is_launchable()
             plan[task] = resources
+            resources_stack.append(resources)
             if len(tasks) == 1:
                 if minimize_cost:
                     objective = sky.Optimizer._compute_total_cost(
                         graph, topo_order, plan)
                 else:
                     objective = sky.Optimizer._compute_total_time(
                         graph, topo_order, plan)
+                if objective < min_objective:
+                    final_plan = {
+                        topo_order[i]: resources_stack[i]
+                        for i in range(len(topo_order))
+                    }
+                    min_objective = objective
             else:
-                objective = _optimize_by_brute_force(tasks[1:], plan)
-            if objective < min_objective:
-                final_plan[task] = resources
-                min_objective = objective
-        return min_objective
+                _optimize_by_brute_force(tasks[1:], plan)
+            resources_stack.pop()
```

Comment:
Say we have a dag: `t1 --> t2`
```python
t1.resources = {r11 (cost=2), r12 (cost=100)}
t2.rersources = {r21 (cost=100), r22 (cost=2)}
```
with egress cost `r12 --> r22 = 1000`, and all other egress cost is 0
Obviously, we have `best_plan = {t1: r11, t2: r22}`, with a total cost of 2 + 2 + 0 (egress) = 4.
But according to the algorithm, when we do `for resources in t1.get_resources():`, it will first try `resources = r11` and set `final_plan = {t1: r11, t2: r22}`;
Later when we try `resources = r12`, let's take a closer look to see what happened:
First try t2 with r21, we get a total cost of 100 + 100 + 0 (egress) = 200;
Then try t2 with r22, we get a total cost of 100 + 2 + 1000 (egress) = 1102.
After this, we will set `final_plan[t2]=r21` (because the object is indeed minimum based on `t1.resources = r12`), and the whole algorithm is finished. So we get `final_plan = {t1: r11, t2: r21}`.

The problem is, we will update `final_plan` only based on the current view, but not a global view; we only find what resources are the best **based on previous recursion choice** and update a **global** dict anyway.

The algorithm to compute `min_objective` is correct though.

Reply:
Ahh, makes sense. Thanks for the detailed explanation. LGTM.